File: picasso-transformations/src/main/java/io/github/hendraanggrian/picassotransformations/color/ColorGrayscaleTransformation.java
package io.github.hendraanggrian.picassotransformations.color;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.support.annotation.NonNull;

import io.github.hendraanggrian.picassotransformations.BaseTransformation;
import io.github.hendraanggrian.picassotransformations.Key;
import io.github.hendraanggrian.picassotransformations.internal.PaintBuilder;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public class ColorGrayscaleTransformation extends BaseTransformation {

    public static final String TAG = "grayscale";

    @NonNull
    @Override
    public Bitmap transform(@NonNull Bitmap source, int height, int width, @NonNull Bitmap target) {
        final ColorMatrix saturation = new ColorMatrix();
        saturation.setSaturation(0f);
        new Canvas(target).drawBitmap(source, 0, 0, new PaintBuilder()
                .colorFilter(new ColorMatrixColorFilter(saturation))
                .build());
        source.recycle();
        return target;
    }

    @NonNull
    @Override
    public Key getKey() {
        return Key.fromTag(TAG);
    }
}

File: picasso-utils/src/com/hendraanggrian/picasso/transformation/ColorGrayscaleTransformation.kt
package com.hendraanggrian.picasso.transformation

import android.graphics.*
import android.graphics.Bitmap.Config.ARGB_8888
import android.graphics.Bitmap.createBitmap
import com.squareup.picasso.Transformation

internal class ColorGrayscaleTransformation : Transformation {

    override fun transform(source: Bitmap): Bitmap {
        val target = createBitmap(source.width, source.height, ARGB_8888)
        Canvas(target).drawBitmap(source, 0f, 0f, Paint().apply {
            colorFilter = ColorMatrixColorFilter(ColorMatrix().apply { setSaturation(0f) })
        })
        source.recycle()
        return target
    }

    override fun key() = "ColorGrayscaleTransformation"
}